### ISSUE TITLE
numpy/core/tests/ should be installed as proper package, not as data

### ISSUE THREAD

**Issue by SoapGentoo** [numpy/numpy#7996]
Dear numpy,
in Gentoo we require python to generate all files (including byte-code `.pyc` and `.pyo`) at build-time and not at runtime, as creating them at runtime creates files that are not tracked by the package manager. Please turn the `numpy/core/tests/` folder into a proper package, such that the `.py` files are compiled during installation.

Gentoo Bugs: https://bugs.gentoo.org/show_bug.cgi?id=590464
Currently we fix it by adding a `setup.py` and `__init__.py` and then tricking the build system into accepting `numpy/core/tests/` as a proper package. In https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=80b2e1b4def977bee9743d7bc52c01acf19fbb9a I do:

```
cp numpy/{compat/setup.py,core/tests}
touch numpy/core/tests/__init__.py
sed \
    -e 's:compat:tests:' \
    -i numpy/core/tests/setup.py
sed \
    -e "s:config\.add_data_dir('tests'):config\.add_subpackage('tests'):" \
    -i numpy/core/setup.py
```

**Comment by rgommers** [comment]
Except it's not a package right? No one is supposed to do ``from numpy.core import tests``, not even inside numpy. Test files are standalone files that are detected and run by nose as standalone files.

**Comment by floppym** [comment]
It seems that *some* code in the wild is importing these modules; otherwise we wouldn't end up with orphaned pyc files on user systems.

**Comment by rgommers** [comment]
running != importing.

what happens is that ``nosetests`` goes looking in subdirs for files named ``test_*``, and then uses those (not sure if it ``exec``s them, parses them or runs them, but it most certainly does not do ``import numpy.core.tests``).

Another way they may get run is directly ``python test_somefile.py`` by a user.

So the issue you're seeing is with some part of the package management tool, I don't think there's a nice way to say "byte-compile these files when installing". Treating them as a subpackage is a hack to achieve that byte-compile step.

**Comment by floppym** [comment]
> So the issue you're seeing is with some part of the package management tool

I don't think so; portage (aka "the package management tool") is not generating the pyc files with unpatched numpy sources. The pyc files are somehow being generated after installation, outside of the control of portage.

Anyway, I'm not sure how the original Gentoo bug reporter managed to end up with pyc files for these test modules in his system, and that's probably worth investigating.

**Comment by rgommers** [comment]
sorry, I meant the Python package management tool (i.e. ``setuptools``, ``pip`` & co), not the Gentoo one.

**Comment by floppym** [comment]
Oh, I see.

Is there any reason to have these test scripts installed on a typical end-user system? Could we prevent them from being installed by setup.py?

**Comment by seberg** [comment]
I would argue that you definitely need to give users the ability to test their install, no matter where it came from. For a distribution install it should never fail, but the end user should be able to check that, especially since it may be possible to use a non-standard setup by using some non-standard blas or so.

**Comment by rgommers** [comment]
I agree, they need to be available. We just need to figure out the best way to byte-compile them during install. Adding ``__init__.py`` is the easiest way to get that done, just not the desired one I think.

**Comment by eric-wieser** [comment]
> No one is supposed to do from numpy.core import tests, 

Well, in python 3 they can anyway

**Comment by rgommers** [comment]
The change to `pytest` has made as add `__init__.py` files in all tests directories. So this now works the way Gentoo wants (I think since numpy 1.15)

**Comment by rgommers** [comment]
Actually, we still use `add_data_dir` rather than `add_subpackage`. Could be changed now, so test files get byte-compiled (that'd be a good thing).

**Comment by seberg** [comment]
Moving the milestone, although it would be nice to do and @hameerabbasi may take a look. It seems easy enough to just change (and make sure all tests are found).

**Comment by charris** [comment]
Fixed by #16173.
